strawberryfields/backends/states.py

```diff
@@ -914,6 +914,38 @@ def number_expectation(self, modes):
             ps = np.tensordot(np.diag(values), ps, axes=((0, 1), (0, 1)))
         return float(ps)
 
+    def diagonal_expectation(self, modes, values):
+        if len(modes) != len(set(modes)):
+            raise ValueError("There can be no duplicates in the modes specified.")
+
+        cutoff = self._cutoff  # Fock space cutoff.
+        num_modes = self._modes  # number of modes in the state.
+
+        traced_modes = tuple(item for item in range(num_modes) if item not in modes)
+        if self.is_pure:
+            # state is a tensor of probability amplitudes
+            ps = np.abs(self.ket()) ** 2
+            ps = ps.sum(axis=traced_modes)
+            for _ in modes:
+                ps = np.tensordot(values, ps, axes=1)
+            return float(ps)
+
+        # state is a tensor of density matrix elements in the SF convention
+        ps = np.real(self.dm())
+        traced_modes = list(traced_modes)
+        traced_modes.sort(reverse=True)
+        for mode in traced_modes:
+            ps = np.tensordot(np.identity(cutoff), ps, axes=((0, 1), (2 * mode, 2 * mode + 1)))
+        for _ in range(len(modes)):
+            ps = np.tensordot(np.diag(values), ps, axes=((0, 1), (0, 1)))
+        return float(ps)
+
+    def parity_expectation(self, modes):
```

Comment:
This looks good. You will also need to write a function for the number expectation, with a very similar logic to this one.

Reply:
Does that mean you want me to delete the old `number_expectation` method?

Reply:
Yes. As far as I can seen you move the functionality into `diagonal_expectation`, now you need to write a wrapper `number_expectation` so that we still have access to that function.